rclcpp/include/rclcpp/copy_all_parameters.hpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 Open Navigation LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__COPY_ALL_PARAMETERS_HPP_
+#define RCLCPP__COPY_ALL_PARAMETERS_HPP_
+
+#include <vector>
+#include <string>
+
+#include "rcl_interfaces/srv/list_parameters.hpp"
+
+#include "rclcpp/parameter.hpp"
+
+namespace rclcpp
+{
+
+/// Copy all parameters from one source node to another destination node.
+/**
+ * \param source Node to copy parameters from
+ * \param destination Node to copy parameters to
+ * \param override_existing_params Default false. Whether to override existing destination params
+ * if both the source and destination contain the same parameter.
+ */
+template<typename NodeT1, typename NodeT2>
+void
+copy_all_parameters(
+  const NodeT1 & source, const NodeT2 & destination, const bool override_existing_params = false)
+{
+  using Parameters = std::vector<rclcpp::Parameter>;
+  auto source_params = source->get_node_parameters_interface();
+  auto dest_params = destination->get_node_parameters_interface();
+
+  std::vector<std::string> param_names = source_params->list_parameters({}, 0).names;
+  Parameters params = source_params->get_parameters(param_names);
+  for (Parameters::const_iterator iter = params.begin(); iter != params.end(); ++iter) {
+    if (!dest_params->has_parameter(iter->get_name())) {
+      dest_params->declare_parameter(iter->get_name(), iter->get_parameter_value());
+    } else if (override_existing_params) {
+      dest_params->set_parameters_atomically({*iter});
```

Comment:
i think this would fail in some cases,

- if the parameter is `read_only`.
- if the parameter value type is different.
- if the value being set does not complys with the descriptor.
- user `on_set_parameters_callbacks` rejects to set the parameter.

after all, with this function, we are not sure what parameters are successfully set to the `destination`. i think that could be complication for user application.

Reply:
I would like to simply remove the `override_existing_params` option then and have it be documented that existing parameters remain untouched. Is that reasonable?

Reply:
i think that your use case is to copy all parameters from `Node` with parameters to `Node` w/o any parameters? in that case, `copy_parameter_declarations` will work. but any other cases, user cannot know how many, what parameter declarations are copied via this method, i think we can describe the current behavior in the doc section.

probably we can consider `override` option for future enhancements.

CC: @wjwwood what do you think?

Reply:
How about `copy_unique_parameters()`? That's a little more niche and implies that it won't copy over shared parameters. If i were in the reviewer's chair, that would meet the _intent_ of my comment but create a specialized case so specialized I'd question the value to `rclcpp` 

I re-read your comment and wondering if you're basically just asking me to error check? 

- If read only --> will that throw? If so, I can catch
- Different values will definitely throw, I can catch
- I'm not really sure what the actionable item is on "if the value being set does not comply with the descriptor". If the parameter is of the same name and type, is that really a realistic concern? If so, how could that be done without machine learning to read the two descriptions to tell if they're meaning the same thing :laughing: 
- Rejection: if the set params API returns to me a status, I can check that

If my reading of this is correct and its just error checking, how yeah, I can do that. I misread it the first go around

Reply:
> I re-read your comment and wondering if you're basically just asking me to error check?

basically yes. these cases happen if it overwrites the parameters via `set_parameter`.
besides, my major concern is user cannot know what parameters are copied successfully or failed, it just works best-effort, and then user needs to check what parameter copied to the other side after all. i am not sure if the application actually rely on this method.

> If the parameter is of the same name and type, is that really a realistic concern?

if the destination parameter is static typed parameter in the description, it cannot copy(set) the different typed value to it. on the other hand, if that is dynamic typed parameter, it can copy and store the different typed value.

https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L263-L306